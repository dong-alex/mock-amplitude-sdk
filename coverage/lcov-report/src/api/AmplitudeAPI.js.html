<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/api/AmplitudeAPI.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">src/api</a> AmplitudeAPI.js
    </h1>
    <div class='clearfix'>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">69x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">69x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">69x</span>
<span class="cline-any cline-yes">52x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">69x</span>
<span class="cline-any cline-yes">72x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">69x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">69x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">5x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">69x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { createContext } from "react";
import useAmplitudeSDK from "../hooks/useAmplitudeSDK";
import generateUUID from "../utils/generateUUID";
&nbsp;
/**
 * Creates a global context within any target application that will allow usage of the mock SDK
 */
export const AmplitudeContext = createContext();
&nbsp;
/**
 * Provider for the amplitude context, this is the access point to all of the functionality that you can
 * use with the mock amplitude SDK. Any adjustments to the SDK and functionality itself will be managed through here.
 * The target application won't be responsible at all for managing this functionality as this will be a seperate library
 * @param {Object} props
 */
export const AmplitudeProvider = props =&gt; {
  const { apiKey, events, initializeAPIKey, logEvent } = useAmplitudeSDK();
&nbsp;
  const handleAPIKeyCreate = apiKey =&gt; {
    initializeAPIKey(apiKey);
  };
&nbsp;
  const handleLogEvent = (eventName, eventProperties) =&gt; {
    logEvent(eventName, eventProperties);
  };
&nbsp;
  const getEvents = () =&gt; {
    return events;
  };
&nbsp;
  const generateAPIKey = <span class="fstat-no" title="function not covered" >()</span> =&gt; {
<span class="cstat-no" title="statement not covered" >    return generateUUID();</span>
  };
&nbsp;
  return (
    &lt;AmplitudeContext.Provider
      value={{
        apiKey,
        handleAPIKeyCreate,
        handleLogEvent,
        getEvents,
        generateAPIKey
      }}
    &gt;
      {props.children}
    &lt;/AmplitudeContext.Provider&gt;
  );
};
&nbsp;
/**
 * Utility wrapper function that will wrap any component in a target application to be given the amplitude functionality for logging events
 * Afterwards, the application will only have to pass in the event name and properties after authentication via the api key which is validated
 * to be true without any sort of data storage for managing valid keys.
 * @param {JSXElement} Component
 */
export const withAmplitudeContext = Component =&gt; {
  return function WrapperComponent(props) {
    return (
      &lt;AmplitudeContext.Consumer&gt;
        {state =&gt; &lt;Component {...props} context={state} /&gt;}
      &lt;/AmplitudeContext.Consumer&gt;
    );
  };
};
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Jan 16 2020 18:29:34 GMT-0700 (Mountain Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
