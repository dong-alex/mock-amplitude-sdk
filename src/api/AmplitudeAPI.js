import React, { createContext } from "react";
import useAmplitudeSDK from "../hooks/useAmplitudeSDK";
import generateUUID from "../utils/generateUUID";

/**
 * Creates a global context within any target application that will allow usage of the mock SDK
 */
export const AmplitudeContext = createContext();

/**
 * Provider for the amplitude context, this is the access point to all of the functionality that you can
 * use with the mock amplitude SDK. Any adjustments to the SDK and functionality itself will be managed through here.
 * The target application won't be responsible at all for managing this functionality as this will be a seperate library
 * @param {Object} props
 */
export const AmplitudeProvider = props => {
  const { apiKey, events, initializeAPIKey, logEvent } = useAmplitudeSDK();

  const handleAPIKeyCreate = apiKey => {
    initializeAPIKey(apiKey);
  };

  const handleLogEvent = (eventName, eventProperties) => {
    logEvent(eventName, eventProperties);
  };

  const getEvents = () => {
    return events;
  };

  const generateAPIKey = () => {
    return generateUUID();
  };

  return (
    <AmplitudeContext.Provider
      value={{
        apiKey,
        handleAPIKeyCreate,
        handleLogEvent,
        getEvents,
        generateAPIKey
      }}
    >
      {props.children}
    </AmplitudeContext.Provider>
  );
};

/**
 * Utility wrapper function that will wrap any component in a target application to be given the amplitude functionality for logging events
 * Afterwards, the application will only have to pass in the event name and properties after authentication via the api key which is validated
 * to be true without any sort of data storage for managing valid keys.
 * @param {JSXElement} Component
 */
export const withAmplitudeContext = Component => {
  return function WrapperComponent(props) {
    return (
      <AmplitudeContext.Consumer>
        {state => <Component {...props} context={state} />}
      </AmplitudeContext.Consumer>
    );
  };
};
