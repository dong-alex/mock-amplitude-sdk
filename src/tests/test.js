import React from "react";
import { render, fireEvent, act, cleanup } from "@testing-library/react";
import { AmplitudeProvider } from "../api/AmplitudeAPI";
import TestComponent from "./TestComponent";

jest.useFakeTimers();

describe("AmplitudeProvider Test", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  afterEach(() => {
    cleanup();
  });

  it("should render a component with AmplitudeProvider information", () => {
    const { container } = render(
      <AmplitudeProvider>
        <TestComponent />
      </AmplitudeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render the message where there is an api key associated", async () => {
    const { getByTestId } = render(
      <AmplitudeProvider>
        <TestComponent />
      </AmplitudeProvider>
    );

    const generateButton = getByTestId("generate-api-button");
    const sessionButton = getByTestId("session-button");

    act(() => {
      fireEvent.click(generateButton);
    });

    act(() => {
      fireEvent.click(sessionButton);
    });

    const apiMessage = getByTestId("message");
    expect(apiMessage.textContent).not.toEqual("");
  });

  it("should flush out events after the time limit of 1 minute has passed", () => {
    const { getByTestId } = render(
      <AmplitudeProvider>
        <TestComponent />
      </AmplitudeProvider>
    );

    const generateButton = getByTestId("generate-api-button");
    const sessionButton = getByTestId("session-button");
    const eventButton = getByTestId("event-button");
    const eventButtonWithProperties = getByTestId("event-button-properties");

    act(() => {
      fireEvent.click(generateButton);
    });

    act(() => {
      fireEvent.click(sessionButton);
    });

    act(() => {
      fireEvent.click(eventButton);
    });

    act(() => {
      fireEvent.click(eventButtonWithProperties);
    });

    act(() => {
      jest.runTimersToTime(60 * 1000);
    });
  });

  it("should load any previous events that were stored locally when the application was last killed", async () => {
    const event1 = new Event("testEvent1");
    const event2 = new Event("testEvent2", { user_id: "11111" });

    localStorage.setItem("amplitude_events", JSON.stringify([event1, event2]));

    const { findAllByTestId } = render(
      <AmplitudeProvider>
        <TestComponent />
      </AmplitudeProvider>
    );

    const allEvents = await findAllByTestId(/log-event-/i);

    expect(allEvents).toHaveLength(2);
  });

  it("should remove all events within the time frame when you exceeded 50 events", () => {
    const { getByTestId } = render(
      <AmplitudeProvider>
        <TestComponent />
      </AmplitudeProvider>
    );

    const generateButton = getByTestId("generate-api-button");
    const sessionButton = getByTestId("session-button");
    const eventButton = getByTestId("event-button");

    act(() => {
      fireEvent.click(generateButton);
    });
    act(() => {
      fireEvent.click(sessionButton);
    });

    for (var i = 0; i < 50; i++) {
      act(() => {
        fireEvent.click(eventButton);
      });
    }
  });
});
