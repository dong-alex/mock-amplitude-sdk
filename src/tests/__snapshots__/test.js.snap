// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmplitudeProvider Test should render a component with AmplitudeProvider information 1`] = `
<div>
  <div
    data-testid="amplitude-ui"
  >
    <label>
      I am an example page
    </label>
    <input
      readonly=""
      value=""
    />
    <button
      data-testid="generate-api-button"
    >
      Generate API Key
    </button>
    <button
      data-testid="session-button"
    >
      Start Amplitude Session
    </button>
    <div
      data-testid="no-api-key-message"
    >
      There is no api-key associated yet
    </div>
    <button
      data-testid="event-button-properties"
    >
      Log Event with properties
    </button>
    <button
      data-testid="event-button"
    >
      Log Event
    </button>
  </div>
</div>
`;
